File: L2/doorman.py
"""https://open.kattis.com/problems/doorman"""
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None
        
    def get(self):
        return self.data
    
    def __str__(self):
        return "(" + str(self.data) + ")"

class ClubLine:
    def __init__(self, head):
        self.head = head
        self.last = head
        
    def __init__(self):
        self.head = None
        self.last = None
    
    def isEmpty(self):
        return self.head is None

    def addLast(self, last):
        node = Node(last)
        if self.head is None:
            self.head = node
            self.last = node
        else:
            self.last.next = node
            tmp = self.last
            self.last = node
            self.last.last = None
            self.last.prev = tmp

    def removeFirst(self):
        if self.head.next is None: # only 1 node
            self.head = None
            self.last = None
        else:
            if self.head.next.next is None: # only 2 nodes
                self.last = self.head.next
            self.head.next.prev = None
            self.head = self.head.next
            
    def hasSecond(self):
        if self.head is None:
            return False
        if self.head.next is None:
            return False
        return True
        
    def removeSecond(self):
        first = self.head
        if first.next.next is not None:
            third = first.next.next
            first.next = third
            third.prev = first
        else:
            first.next = None
            self.last = first

    def __str__(self):
        if self.head is None:
            return "()"
        to_return = str(self.head)
        cur = self.head
        while cur.next is not None:
            to_return += " -> " + str(cur.next)
            cur = cur.next
        return to_return
            
if __name__ == "__main__":
    iq = int(input())
    q = input()
    line = ClubLine()
    for human in range(len(q)):
        line.addLast(q[human])
    diff = 0
    entered = 0                      # total humans entered in the club
    while abs(diff) <= iq:
        if line.isEmpty():
            break
        
        if abs(diff) < iq:
            if line.head.get() == "W":
                diff -= 1
            if line.head.get() == "M":
                diff += 1
            entered += 1
            line.removeFirst()
            
        elif abs(diff) == iq:
            if diff > 0:
                if line.head.get() == "W":
                    diff -= 1
                    entered += 1
                    line.removeFirst()
                elif line.head.next.get() == "W":
                    diff -= 1
                    entered += 1
                    line.removeSecond()
                else:
                    break
            elif diff < 0:
                if line.head.get() == "M":
                    diff += 1
                    entered += 1
                    line.removeFirst()
                elif line.head.next.get() == "M":
                    diff += 1
                    entered += 1
                    line.removeSecond()
                else:
                    break
        else:
            break
    print(entered)

File: L2/doorman.go
// https://open.kattis.com/problems/doorman
package main

import (
	"fmt"
)

type Node struct {
	prev *Node
	next *Node
	data string
}

type List struct {
	head *Node
	tail *Node
}

func (node *Node) GetData() string {
	return node.data
}

func (club *List) Append(data string) {		// append to last
	human := &Node{
		data: data,
	}
	if club.head == nil {
		club.head = human
		club.tail = human
	} else {
		club.tail.next = human
		tmp := club.tail
		club.tail = human
		club.tail.prev = tmp
		club.tail.next = nil
	}
}

func (club *List) RemoveFirst() {
	 if club.head.next == nil {				// 1 node
		club.head = nil
		club.tail = nil
	} else if club.head.next.next == nil {	// 2 nodes
		club.tail = club.head.next
		club.head = club.head.next
	} else {
		club.head = club.head.next
		club.head.next.prev = nil
	}
}

func (club *List) HasSecond() bool {			// check if has at least 2 nodes
	if club.head == nil || club.head.next == nil {
		return false
	}
	return true
}

func (club *List) RemoveSecond() {
	if club.head.next.next != nil {
		club.head.next = club.head.next.next // first is now pointed to third instead of second
		club.head.next.next.prev = club.head.next.next
	} else {
		club.head.next = nil				// nothing to point anymore
		club.tail = club.head
	}
}

func (club *List) Display() {
	if club.head == nil {
		fmt.Println("( )")
	} else {
		return_list := club.head
		toReturn := ""
		for return_list.next != nil {
			toReturn += "(" + return_list.data + ") -> "
			return_list = return_list.next
		}
		toReturn += "(" + return_list.data + ")"
		fmt.Println(toReturn)
	}
}

func Abs(n int) int {
	if n > 0 {
		return n
	} else {
		return -n
	}
}

func main() {
	var club List
	var smart int
	var queue string

	fmt.Scan(&smart)
	fmt.Scan(&queue)
	for _, human := range queue {
		club.Append(string(human))
	}
	diff := 0
	entered := 0
	for Abs(diff) <= smart {
		if club.head == nil { 		// club is empty
			break
		}
		if Abs(diff) < smart {
			entered++
			if club.head.GetData() == "M" {
				diff++
			} else {
				diff--
			}
			club.RemoveFirst()
		} else {
			if diff > 0 {			// too many men
				if club.head.GetData() == "W" {
					diff--
					entered++
					club.RemoveFirst()
				} else if club.head.next.GetData() == "W" {
					diff--
					entered++
					club.RemoveSecond()
				} else {
					break
				}
			} else {				// too many women
				if club.head.GetData() == "M" {
					diff++
					entered++
					club.RemoveFirst()
				} else if club.head.next.GetData() == "M" {
					diff++
					entered++
					club.RemoveSecond()
				} else {
					break
				}
			}
		}
	}
	fmt.Print(entered)
}